packages/langium-cli/src/generator/highlighting/monarch-generator.ts

```diff
@@ -0,0 +1,476 @@
+        genLanguageDefEntry('keywords', languageDef.keywords, indent),
+        genLanguageDefEntry('operators', languageDef.operators, indent),
+        // special case, identify symbols via singular regex
+        indent('symbols:  /' +  languageDef.symbols.join('|') + '/,')
+    ];
+    return content.join('\n') + '\n';
+}
+
+/**
+ * Pretty prints the tokenizer portion of a Monarch grammar file
+ * @param tokenizer Tokenizer portion to print out
+ * @returns Tokenizer in concrete form
+ */
+function prettyPrintTokenizer(tokenizer: Tokenizer): string {
+    return [
+        '\ttokenizer: {',
+        tokenizer.states.map(state => prettyPrintState(state, (s => indent(indent(s))))).join('\n'),
+        '\t}'
+    ].join('\n');
+}
+
+/**
+ * Pretty prints a tokenizer state, composed of various rules
+ * @param state Tokenizer state to pretty print
+ * @param fmt Formatter to set indentation
+ * @returns Tokenizer state in concrete form
+ */
+function prettyPrintState(state: State, fmt: Formatter): string {
+    return [
+        fmt(state.name + ': ['),
+        state.rules.map(rule => fmt(prettyPrintRule(rule, indent))).join('\n'),
+        fmt('],')
+    ].join('\n');
+}
+
+/**
+ * Pretty prints a Rule.
+ * This can either be a literal rule to match w/ an action, or a reference to a state to include here
+ * @param ruleOrState Rule to pretty print. If it's a state, we include that state's contents implicitly within this context.
+ * @param fmt Formatter to track indentation
+ * @returns Rule in concrete form
+ */
+function prettyPrintRule(ruleOrState: Rule | State, fmt: Formatter): string {
+    if(isRule(ruleOrState)) {
+        // extract rule pattern, either just a string or a regex w/ parts
+        const rulePatt = ruleOrState.regex instanceof RegExp ? getTerminalParts(ruleOrState.regex).join('') : `/${ruleOrState.regex}/`;
+        return fmt('{ regex: ' + rulePatt + ', action: ' + prettyPrintAction(ruleOrState.action) + ' },');
+    } else {
+        // include another state by name, implicitly includes all of its contents
+        return fmt(`{ include: '@${ruleOrState.name}' },`);
+    }
+}
+
+/**
+ * Pretty prints the action of a Rule
+ * @param action Action to print. Can have several keywords to control what the state machine should do next.
+ * @returns Action in concrete form
+ */
+function prettyPrintAction(action: Action | Case[]): string {
+    if(!Array.isArray(action)) {
+        // plain action
+        return JSON.stringify(action);
+    } else {
+        // array of cases, each with an action
+        const prettyCases: string = action.map(c => `'${c.guard}': ` + prettyPrintAction(c.action)).join(', ');
+        return '{ cases: { ' + prettyCases + ' }}';
+    }
+}
+
+/**
+ * Convert a deafult Langium token names to a monarch one
+ * @param name Token name to convert
+ * @returns Returns the equivalent monarch name, or the original token name
+ */
+function getMonarchTokenName(name: string): string {
```

Comment:
Suggestion: These names are quite specific. We perform a more sophisticated analysis of what a comment or string is in the texmate Generator. Especially when it comes to `STRING`, `INT` and `Bigint` we can simply look at the return types of the terminal rules.

Reply:
Seems we just check `rule.name` for string (lower cased) in TM. I've done the same here, using the type name when possible, and falling back on the rule name when that isn't available.